src/GafferScene/PathFilter.cpp

```diff
@@ -166,17 +234,110 @@ void PathFilter::hashMatch( const ScenePlug *scene, const Gaffer::Context *conte
 	{
 		pathMatcherPlug()->hash( h );
 	}
+
+	if( rootsPlug()->getInput() )
+	{
+		rootSizesPlug()->hash( h );
+	}
 }
 
 unsigned PathFilter::computeMatch( const ScenePlug *scene, const Gaffer::Context *context ) const
 {
 	const ScenePlug::ScenePath &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
 	ConstPathMatcherDataPtr pathMatcher = m_pathMatcher ? m_pathMatcher : pathMatcherPlug()->getValue();
-	return pathMatcher->readable().match( path );
+
+	if( !rootsPlug()->getInput() )
+	{
+		return pathMatcher->readable().match( path );
+	}
+
+	ConstIntVectorDataPtr rootSizes = rootSizesPlug()->getValue();
+	ScenePlug::ScenePath relativePath = path;
+	unsigned result = PathMatcher::NoMatch;
+
+	size_t previousRootSize = 0;
+	for( const int rootSize : rootSizes->readable() )
+	{
+		if( rootSize == g_descendantRoots )
+		{
+			if( !pathMatcher->readable().isEmpty() )
+			{
+				result |= PathMatcher::DescendantMatch;
+			}
+			break;
+		}
+		relativePath.erase( relativePath.begin(), relativePath.begin() + rootSize - previousRootSize );
+		result |= pathMatcher->readable().match( relativePath );
+		previousRootSize = rootSize;
+	}
+
+	return result;
+}
+
+bool PathFilter::sceneAffectsMatch( const ScenePlug *scene, const Gaffer::ValuePlug *child ) const
+{
+	if( Filter::sceneAffectsMatch( scene, child ) )
+	{
+		return true;
+	}
+
+	return rootsPlug()->sceneAffectsMatch( scene, child );
 }
 
-		ConstPathMatcherDataPtr pathMatcher = m_pathMatcher ? m_pathMatcher : pathMatcherPlug()->getValue();
-		return pathMatcher->readable().match( pathData->readable() );
+void PathFilter::hashRootSizes( const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	const ScenePlug::ScenePath &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+	if( path.size() )
+	{
+		ScenePlug::PathScope parentScope( context, ScenePlug::ScenePath( path.begin(), path.begin() + path.size() - 1 ) );
+		rootSizesPlug()->hash( h );
 	}
-	return IECore::PathMatcher::NoMatch;
+	rootsPlug()->hash( h );
+	h.append( (uint64_t)path.size() );
+}
+
+ConstIntVectorDataPtr PathFilter::computeRootSizes( const Gaffer::Context *context ) const
+{
+	const ScenePlug::ScenePath &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+
+	// Start with the root sizes from our parent.
+
+	ConstIntVectorDataPtr parentRootSizes;
+	if( path.size() )
+	{
+		ScenePlug::PathScope parentScope( context, ScenePlug::ScenePath( path.begin(), path.begin() + path.size() - 1 ) );
+		parentRootSizes = rootSizesPlug()->getValue();
+		// If the parent has no descendant roots, then we already have
+		// all the roots we need.
+		if( !parentRootSizes->readable().size() || parentRootSizes->readable().back() != g_descendantRoots )
+		{
+			return parentRootSizes;
+		}
+	}
+
+	// Then figure out if there is a new root here,
+	// and if there are still any descendant roots.
+
+	const unsigned m = rootsPlug()->getValue();
```

Comment:
Are we happy we're already in a `FilterPlug::SceneScope`?

Reply:
Yes. We only pull on `rootSizesPlug()` from `hashMatch()/computeMatch()` and `hashRootSizes()/computeRootSizes()`. Both `hashMatch()` and `computeMatch()` require that we have a `SceneScope` already.